### PR TITLE
Allow dynamic types in public mappings

### LINKED ISSUE #498: Implement accessors for mappings with strings as keys

### PR DESCRIPTION
**Pull request by Marenz**
fixes #498

### CODE REVIEW

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 1]
```diff
@@ -106,18 +106,38 @@ void ExpressionCompiler::appendStateVariableAccessor(VariableDeclaration const&
...
 		{
 			solAssert(CompilerUtils::freeMemoryPointer >= 0x40, "");
-			solUnimplementedAssert(
-				!paramTypes[i]->isDynamicallySized(),
-				"Accessors for mapping with dynamically-sized keys not yet implemented."
-			);
+
 			// pop offset
 			m_context << Instruction::POP;
 			// move storage offset to memory.
```
Comment (chriseth) [inline libsolidity/codegen/ExpressionCompiler.cpp]: Please move this comment down

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 2]
```diff
@@ -106,18 +106,38 @@ void ExpressionCompiler::appendStateVariableAccessor(VariableDeclaration const&
...
-			utils().storeInMemory(32);
-			// move key to memory.
-			utils().copyToStackTop(paramTypes.size() - i, 1);
-			utils().storeInMemory(0);
-			m_context << u256(64) << u256(0) << Instruction::KECCAK256;
+			if (paramTypes[i]->isDynamicallySized())
+			{
+				utils().copyToStackTop(paramTypes.size() - i + 1, 1);
```
Comment (chriseth) [inline libsolidity/codegen/ExpressionCompiler.cpp:122]: Please add a comment about what this copies to stack top.

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 3]
```diff
@@ -106,18 +106,38 @@ void ExpressionCompiler::appendStateVariableAccessor(VariableDeclaration const&
...
+
+				utils().fetchFreeMemoryPointer();
+				// stack: base index mem
+				// note: the following operations must not allocate memory!
```
Comment (chriseth) [inline libsolidity/codegen/ExpressionCompiler.cpp]: Is this really a requirement here?

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 4]
```diff
@@ -106,18 +106,38 @@ void ExpressionCompiler::appendStateVariableAccessor(VariableDeclaration const&
...
+					TypePointers{paramTypes[i]},
+					TypePointers{paramTypes[i]}
+				);
+
+				// Store slot position in memory after key data
+				m_context << Instruction::SWAP1;
+				utils().storeInMemoryDynamic(IntegerType::uint256());
+
+				utils().toSizeAfterFreeMemoryPointer();
+			}
```
Comment (chriseth) [inline libsolidity/codegen/ExpressionCompiler.cpp:137]: What is the stack layout here?

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 5]
```diff
@@ -106,18 +106,38 @@ void ExpressionCompiler::appendStateVariableAccessor(VariableDeclaration const&
...
+
+				utils().fetchFreeMemoryPointer();
+				// stack: base index mem
+				// note: the following operations must not allocate memory!
+				utils().packedEncode(
+					TypePointers{paramTypes[i]},
+					TypePointers{paramTypes[i]}
+				);
+
+				// Store slot position in memory after key data
```
Comment (chriseth) [inline libsolidity/codegen/ExpressionCompiler.cpp]: Would also be good to know the stack layout here.

File: test/libsolidity/SolidityEndToEndTest.cpp [hunk 6]
```diff
@@ -8435,6 +8435,75 @@ BOOST_AUTO_TEST_CASE(string_as_mapping_key)
...
+	vector<string> strings{
+		"Hello, World!",
+		"Hello,                            World!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!1111",
+		"",
+		"1"
+	};
+	for (unsigned i = 0; i < strings.size(); i++)
+		ABI_CHECK(callContractFunction(
+			"set(string,string,uint256)",
+			u256(0x60),
```
Comment (chriseth) [inline test/libsolidity/SolidityEndToEndTest.cpp:8495]: Please use two different strings. If the encoding is too cumbersome, you can always use Solidity code that calls `this.data("abc", "def")`.

**Review by chriseth — CHANGES_REQUESTED** [review]
Please also add tests that mix strings and value types as keys.
Furthermore, please verify that it also works correctly for other dynamic types like nested dynamic arrays.

File: libsolidity/codegen/ExpressionCompiler.cpp (hunk 3 above)
Comment (Marenz) [reply libsolidity/codegen/ExpressionCompiler.cpp]: well, I suppose calling a function should generally not allocate memory?

Comment (chriseth) [reply libsolidity/codegen/ExpressionCompiler.cpp]: It is a different case inside user code or whether you have full control over everything.

**Comment by Marenz** [comment]
Pushed another change.
After talking a bit with Chris, we discovered that we don't actually need to copy the key data around. Instead we use it in place and temporarily use the 32 bit after the key data for the slot

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 7]
```diff
@@ -106,18 +106,75 @@ void ExpressionCompiler::appendStateVariableAccessor(VariableDeclaration const&
...
-			m_context << u256(64) << u256(0) << Instruction::KECCAK256;
+			if (paramTypes[i]->isDynamicallySized())
+			{
+				// stack: <keys..> <slot position>
+
+				// copy key[i] to top.
+				utils().copyToStackTop(paramTypes.size() - i + 1, 1);
+
+				// stack: <keys..> <slot position> <key1>
+				// Load key length to stack
```
Comment (chriseth) [inline libsolidity/codegen/ExpressionCompiler.cpp]: You can only use the shortcut if the type is actually a `string` or `bytes` type - please see the KECCAK256 function implementation in ExpressionCompiler. I would also advise to use `m_context.appendInlineAssembly(...)` here.

File: test/libsolidity/SolidityEndToEndTest.cpp [hunk 8]
```diff
@@ -51,6 +51,11 @@ namespace test
 
 BOOST_FIXTURE_TEST_SUITE(SolidityEndToEndTest, SolidityExecutionFramework)
 
+int roundTo32 ( int _num )
+{
+	return ceil(_num / 32.0) * 32;
```
Comment (axic) [inline test/libsolidity/SolidityEndToEndTest.cpp]: Could just use `(num + 31) / 32`.

Also could make it `constexpr`.

Comment (chriseth) [reply test/libsolidity/SolidityEndToEndTest.cpp]: and it should probably go into devcore, turned into a template and named something like `smallestMultipleOf32`.

Comment (Marenz) [reply test/libsolidity/SolidityEndToEndTest.cpp]: And another two things learned

File: libsolidity/codegen/ExpressionCompiler.cpp (hunk 7 above)
Comment (Marenz) [reply libsolidity/codegen/ExpressionCompiler.cpp]: But according to the documentation `string` and `bytes` are the only dynamic types allowed for mappings?

File: test/libsolidity/SolidityEndToEndTest.cpp [hunk 9]
```diff
@@ -51,6 +51,11 @@ namespace test
 
 BOOST_FIXTURE_TEST_SUITE(SolidityEndToEndTest, SolidityExecutionFramework)
 
+int constexpr roundTo32 ( int _num )
```
Comment (chriseth) [inline test/libsolidity/SolidityEndToEndTest.cpp]:
```suggestion
int constexpr roundTo32(int _num)
```

File: test/libsolidity/SolidityEndToEndTest.cpp [hunk 10]
```diff
@@ -8435,6 +8440,142 @@ BOOST_AUTO_TEST_CASE(string_as_mapping_key)
 		), encodeArgs(u256(7 + i)));
 }
 
+BOOST_AUTO_TEST_CASE(string_as_public_mapping_key)
+{
+	char const* sourceCode = R"(
+		contract Test { mapping(string => uint) public data;
```
Comment (chriseth) [inline test/libsolidity/SolidityEndToEndTest.cpp]:
```suggestion
		contract Test {
		    mapping(string => uint) public data;
```

File: test/libsolidity/SolidityEndToEndTest.cpp [hunk 11]
```diff
@@ -8435,6 +8440,142 @@ BOOST_AUTO_TEST_CASE(string_as_mapping_key)
...
+			u256(0x20),
+			u256(strings[i].size()),
+			strings[i]
+		), encodeArgs(u256(7 + i)));
+}
+
+BOOST_AUTO_TEST_CASE(nested_string_as_public_mapping_key)
+{
+	char const* sourceCode = R"(
+		contract Test { mapping(string => mapping(string => uint)) public data;
```
Comment (chriseth) [inline test/libsolidity/SolidityEndToEndTest.cpp]:
```suggestion
		contract Test {
		    mapping(string => mapping(string => uint)) public data;
```

**Review by chriseth — APPROVED** [review]

**Pull request merged**